gateway/src/main/java/io/camunda/zeebe/gateway/admin/backup/BackupRequestHandler.java

```diff
@@ -25,40 +31,144 @@ public BackupRequestHandler(final BrokerClient brokerClient) {
+            .toList();
+
+    return CompletableFuture.allOf(statusesReceived.toArray(CompletableFuture[]::new))
+        .thenApply(ignore -> aggregatePartitionStatus(backupId, statusesReceived))
+        .exceptionallyCompose(
+            error ->
+                CompletableFuture.failedFuture(
+                    operationFailed("query", backupId, error.getCause())));
+  }
+
+  private Either<Throwable, Boolean> checkTopologyComplete() {
+    final BrokerClusterState topology = topologyManager.getTopology();
+    if (topology == null) {
+      return Either.left(new NoTopologyAvailableException());
+    }
+
+    final int expectedPartitionCount = topology.getPartitionsCount();
+    final int knownPartitions = topology.getPartitions().size();
+    if (expectedPartitionCount != knownPartitions) {
+      return Either.left(
+          new IncompleteTopologyException(
+              "Expected to send request to all %d partitions, but found only %d partitions in topology."
+                  .formatted(expectedPartitionCount, knownPartitions)));
+    }
+
+    return Either.right(true);
+  }
+
+  private static BackupOperationFailedException operationFailed(
+      final String operation, final long backupId, final Throwable error) {
+    return new BackupOperationFailedException(operation, backupId, error);
+  }
+
+  private BackupStatus aggregatePartitionStatus(
+      final long backupId,
+      final List<CompletableFuture<BrokerResponse<BackupStatusResponse>>> completedFutures) {
+
+    final var partitionStatuses =
+        completedFutures.stream()
+            .map(response -> response.join().getResponse())
+            .map(PartitionBackupStatus::from)
+            .toList();
+
+    final var combinedStatus = getAggregatedStatus(partitionStatuses);
+
+    if (combinedStatus.isEmpty()) {
+      throw new IllegalStateException(
+          "Backup status cannot be calculated from %s".formatted(partitionStatuses));
+    }
+
+    String failureReason = null;
+    if (combinedStatus.get() == BackupStatusCode.FAILED) {
+      failureReason = collectFailureReason(partitionStatuses);
+    }
+    return new BackupStatus(
+        backupId, combinedStatus.get(), Optional.ofNullable(failureReason), partitionStatuses);
   }
 
-  private static BackupFailedException backupFailed(final long backupId, final Throwable error) {
-    return new BackupFailedException(backupId, error);
+  private String collectFailureReason(final List<PartitionBackupStatus> partitionStatuses) {
+    return partitionStatuses.stream()
+        .filter(p -> p.status() == BackupStatusCode.FAILED)
+        .map(
+            p -> {
+              final var reason = p.failureReason().orElse("Unknown reason");
+              return "Backup on partition %d failed due to %s.".formatted(p.partitionId(), reason);
+            })
+        .toList()
+        .toString();
+  }
+
+  private static Optional<BackupStatusCode> getAggregatedStatus(
```

Comment:
:question: It's unclear to me why both this and `combine` are static, as they're only used in this class, in an instance context. Is this a performance optimization?

Reply:
No need to apologize, was just wondering :smile: